contributors/design-proposals/node/node-usernamespace-remapping.md

```diff
@@ -48,134 +49,248 @@ Enable user namespace support in a kubernetes cluster so that workloads that wor
+
+
+## Implementation Roadmap
+### Phase 1: Support in Kubelet, Alpha, [Target: Kubernetes v1.14]
+- Add `hostUserNamespace` in pod API
+- Add feature gate `HostUserNamespace`, disabled by default
 - Add new CRI API, `GetRuntimeConfigInfo()`
 - Add logic in Kubelet to handle pod creation which includes parsing GetRuntimeConfigInfo response and changing file-permissions in /var/lib/kubelet with learned userns mapping.
 - Add changes in dockershim to implement GetRuntimeConfigInfo() for docker runtime
 - Add changes in CRI-O to implement userns support and GetRuntimeConfigInfo() support
 - Unit test cases
 - e2e tests
 
-#### Phase 2: Beta Support [Target: Kubernetes v1.12]
+### Phase 2: Beta Support [Target: Kubernetes v1.15]
 - PSP integration
-- To grow ExperimentalHostUserNamespaceDefaulting from experimental feature gate to a Kubelet flag
-- API changes to allow pod able to request HostUserNamespace in pod spec
-- e2e tests
+- Namespace-Level usernamespace remapping support
 
-### References
+## Rollout Strategy
+Upgrading a cluster to 1.14 will not impact any existing workloads (Assuming that existing cluster nodes were not running runtimes with usernamespace remapping enabled). To be able to use the security benefit of usernamespace remapping on existing node, usernamespace remapping will have to enabled at the runtime. To perform this runtime configuration update, first node will be tainted to make it unschedulable, then it will be drained and runtime will be updated. Next, the taint will removed to make it available to the schedular. Now Kubelet will discover this configuration change and starts creating pods with remapped usernamespace.
+As we discussed in above sections, Values of `RunAsUser`, `RunAsGroup`, `SupplementalGroups` and `FsGroup` in the older/existing templates may cause pod admission failure if the values are not consistent with the default id mappings of the runtime as learned by Kubelet.
+
+`RunAsUser` and `RunAsGroup` values are treated as translated IDs in the container usernamespace and if the runtime's default mappings does not have a mapping for these containerIds, pod will fail admission at Kubelet.
+
+`FsGroup` and `SupplementalGroups` values are are not treated as translated IDs. For a FsGroup/SupplementalGroup ID, container usernamespace ID is mapped to the same host usernamespace ID. Therefore for these GIDs in the pod objects in the older templates, either there should be no mapping in runtime's default id mappings or if mapping is there, container id and host id should be same fopr the mapping range i.e 1:1 mapping.
+
+### Recommended default id mappings configuration for the runtime
+- Map `0` {U/G}ID in the container namespace with max {u/g}id i.e `4294967295`
+- Map `1` ID to `4294967294`(max id -1) ID in the container namespace with the exactly same IDs on host user namespace
```

Comment:
sounds good, but then any container using supplemental or fsgroups will fail. will the kubelet detect that no translation is needed, and not cause problems?

Reply:
yeah, either kubelet can detect and will not set supplemental groups in the pod creation request which is sent to the runtime.
OR
Can runtimes be expected to understand this and behave properly?

I think first is more sane. WDYT?